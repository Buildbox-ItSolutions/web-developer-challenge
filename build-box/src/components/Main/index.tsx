import './Main.css';
import Form from '../Form';
import Post from '../Post';
import posts from '../../assets/data/posts';

function Main() {
	return (
		<main>
			<Form />
			<section>
				<h3>Feed</h3>
				{posts.map(({ image, message, author, id }) => (
					<Post
						image={image}
						message={message}
						author={author}
						key={id}
						uuid={id}
					/>
				))}
			</section>
		</main>
	);
}

export default Main;
