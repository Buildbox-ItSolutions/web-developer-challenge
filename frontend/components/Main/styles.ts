import styled from 'styled-components';

export const Main = styled.main`
    padding-top: 20px;
    width: 100%;
    height: auto;
    grid-area: main;
    background-color: #343434;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: center;

    @media (min-width: 900px) {
        display: grid;

        grid-template-rows: auto auto auto;

        grid-template-columns: 1.5fr 2fr 1.5fr;
    }
`;

export const Feed = styled.p`
    width: 100%;
    height: 100px;
    color: #7a7a7a;
    font-size: 14px;
    grid-column-start: 2;
    grid-column-end: 2;
    display: grid;
    align-items: flex-end;
    margin-bottom: 10px;

    @media (max-width: 900px) {
        display: flex;
        justify-content: center;
        align-items: center;
    }
`;
