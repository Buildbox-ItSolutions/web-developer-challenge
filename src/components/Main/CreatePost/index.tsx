/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @next/next/no-img-element */

import { useContext, useRef, useState } from 'react'
import { v4 as uuidv4 } from 'uuid'
import { ToastContainer, toast } from 'react-toastify'
import 'react-toastify/dist/ReactToastify.css'

import PostsContext from '../../../shared/contexts/posts'
import { getBase64 } from '../../../shared/utils/getBase64'
import { CreatePostContainer } from './style'

export default function CreatePost() {
  const { posts, setPosts } = useContext(PostsContext)

  const [base64, setBase64] = useState<string>('')

  const inputNameRef = useRef<HTMLInputElement>(null)
  const inputMessageRef = useRef<HTMLTextAreaElement>(null)

  function cleanForm() {
    if (inputNameRef.current && inputMessageRef.current) {
      setBase64('')
      inputNameRef.current.value = ''
      inputMessageRef.current.value = ''
    }
  }

  async function handleSubmit(event: React.FormEvent) {
    event.preventDefault()

    if (
      inputNameRef.current &&
      inputMessageRef.current &&
      inputNameRef.current.value !== '' &&
      inputMessageRef.current.value !== '' &&
      base64 !== ''
    ) {
      const newPost = {
        id: uuidv4(),
        image: base64,
        name: inputNameRef.current.value,
        message: inputMessageRef.current.value,
      }

      setPosts([...posts, newPost])

      cleanForm()

      toast.success('Publicação enviada com sucesso!', {
        position: toast.POSITION.TOP_CENTER,
        autoClose: 2000,
      })
    } else {
      if (base64 === '') {
        toast.error('Adicione uma imagem à publicação!', {
          position: toast.POSITION.TOP_CENTER,
          autoClose: 2000,
        })
      }
      if (inputNameRef.current && inputNameRef.current.value === '') {
        toast.error('Preencha seu nome na publicação!', {
          position: toast.POSITION.TOP_CENTER,
          autoClose: 2000,
        })
      }

      if (inputMessageRef.current && inputMessageRef.current.value === '') {
        toast.error('Escreva uma mensagem na publicação!', {
          position: toast.POSITION.TOP_CENTER,
          autoClose: 2000,
        })
      }
    }
  }

  async function onChangeFileInput(file: File) {
    const newBase64 = await getBase64(file)

    setBase64(newBase64 as string)
  }

  return (
    <CreatePostContainer>
      <div className="card">
        <form
          onSubmit={(event) => {
            handleSubmit(event)
          }}
        >
          <div className="selectImage">
            {base64 === '' ? (
              <label htmlFor="imageFileInput">
                <img src="/image.png" alt="image icon" />
              </label>
            ) : (
              <label htmlFor="imageFileInput">
                <img
                  src={base64}
                  alt="selected image"
                  className="selectedImage"
                />
              </label>
            )}
            <input
              id="imageFileInput"
              type="file"
              onClick={(e) => ((e.target as HTMLInputElement).value = '')}
              onChange={(e) => {
                if (e.target.files) {
                  onChangeFileInput(e.target.files[0])
                }
              }}
              accept=".png, .jpg, .jpeg"
            />
            {base64 !== '' ? (
              <img
                src="/trash.png"
                alt="trash icon"
                className="trashIcon"
                onClick={() => setBase64('')}
              />
            ) : (
              ''
            )}
          </div>
          <input
            className="inputName"
            ref={inputNameRef}
            type="text"
            name="name"
            placeholder="Digite seu nome"
          />
          <textarea ref={inputMessageRef} placeholder="Mensagem" />
          <div className="buttonsArea">
            <button className="descartar" onClick={cleanForm} type="button">
              Descartar
            </button>
            <button className="publicar" type="submit">
              Publicar
            </button>
          </div>
        </form>
      </div>
      <ToastContainer limit={3} />
    </CreatePostContainer>
  )
}
