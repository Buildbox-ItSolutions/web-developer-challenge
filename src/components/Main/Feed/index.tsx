import { useContext } from 'react'

import PostsContext from '../../../shared/contexts/posts'
import Spinner from '../../Spinner'
import PostCard from '../PostCard'
import { FeedContainer } from './style'

export default function Feed() {
  const { posts } = useContext(PostsContext)

  if (!posts) {
    return (
      <FeedContainer>
        <div className="titleArea">
          <h1>Feed</h1>
        </div>
        <div style={{ margin: '2rem 0' }}>
          <Spinner />
        </div>
      </FeedContainer>
    )
  } else {
    return (
      <FeedContainer>
        <div className="titleArea">
          <h1 className="title">Feed</h1>
        </div>
        {posts.length === 0 ? (
          <h1 className="noContent">Nenhuma publicação encontrada...</h1>
        ) : (
          posts.map((post) => {
            return <PostCard key={post.id} post={post} />
          })
        )}
      </FeedContainer>
    )
  }
}
