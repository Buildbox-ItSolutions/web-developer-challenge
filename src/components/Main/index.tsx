import React from "react";
import { StyledMain } from "./style";

interface MainProps {
  children: React.ReactNode;
}

export const Main = ({ children }: MainProps) => {
  return <StyledMain>{children}</StyledMain>;
};
