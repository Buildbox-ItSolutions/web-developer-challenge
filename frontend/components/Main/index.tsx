import { FC, FormEvent, useState } from 'react';
import { v4 as uuidv4 } from 'uuid';

import Card from '../Card';
import Publisher from '../Publisher';
import { Feed, Main } from './styles';

interface Post {
    author: string;
    profilePic: string;
    message: string;
    id?: string;
}

export interface submitProps {
    author: string;
    profilePic: string;
    message: string;
    e: FormEvent<HTMLFormElement>;
}

const MainComponent: FC = () => {
    const [posts, setPosts] = useState<Post[]>([]);

    const handleSubmit = (props: submitProps) => {
        props.e.preventDefault();
        const id = uuidv4();

        setPosts((prev) => [
            ...prev,
            {
                author: props.author,
                message: props.message,
                profilePic: props.profilePic || '/default-profile.png',
                id
            }
        ]);
    };

    const hendleDelete = (id: string) => {
        const updatedPosts = posts.filter((post) => post.id !== id);

        setPosts(updatedPosts);
    };
    return (
        <Main>
            <Publisher handleSubmit={handleSubmit} data-test="publisher-component" />
            <Feed>Feed</Feed>

            {posts &&
                posts.map((post) => (
                    <Card
                        hendleDelete={hendleDelete}
                        key={post.id}
                        profilePic={post.profilePic}
                        autor={post.author}
                        message={post.message}
                        id={post.id}
                    />
                ))}
        </Main>
    );
};

export default MainComponent;
