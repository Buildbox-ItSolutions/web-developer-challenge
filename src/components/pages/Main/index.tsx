import React from 'react';
import { useSelector } from 'react-redux';

import { Header } from 'components/atoms';
import { CreatePost, PostItem } from 'components/molecules';

import { RootState } from 'store/ducks/rootReducer';
import { Container } from './styles';

const Main: React.FC = () => {
  const posts = Array.from(useSelector((state: RootState) => state.posts));

  return (
    <>
      <Container>
        <Header />
        <main>
          <CreatePost />
          {posts.map(post => (
            <PostItem postData={post} key={post.id} />
          ))}
        </main>
      </Container>
    </>
  );
};

export default Main;
