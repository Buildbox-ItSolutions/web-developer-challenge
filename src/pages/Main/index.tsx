import { Header } from '../../components/Header';
import { CreatePost } from '../../features/posts/CreatePost';
import { PostItem } from '../../features/posts/PostItem';
import { IPost } from '../../models/IPost';
import { useAppSelector } from '../../app/hooks';

import { Container } from './styles';
import { selectPosts } from '../../features/posts/postsSlice';

export function Main() {
  const posts: IPost[] = useAppSelector(selectPosts);

  return (
    <>
      <Container>
        <Header />
        <main>
          <CreatePost />
          {posts.map(post => (
            <PostItem postData={post} key={post.id} />
          ))}
        </main>
      </Container>
    </>
  );
};
