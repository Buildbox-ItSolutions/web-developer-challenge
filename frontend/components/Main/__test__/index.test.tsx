import 'jest-styled-components';

import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import Main from '../index';

describe('Main', () => {
    it('must render without errors', () => {
        const wrapper = renderer.create(<Main />).toJSON();

        expect(wrapper).toMatchSnapshot();
    });

    it('must display a publisher component ', () => {
        const wrapper = shallow(<Main />);

        expect(wrapper.find(`[data-test='publisher-component']`).exists()).toBeTruthy();
    });
});
