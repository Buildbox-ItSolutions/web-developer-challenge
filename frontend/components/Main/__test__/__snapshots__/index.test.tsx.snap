// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main must render without errors 1`] = `
.c1 {
  width: 80%;
  height: 353px;
  background-color: #313131;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 3px;
  padding: 2em;
  border: 1px solid #3b3b3b;
  grid-column-start: 2;
  grid-column-end: 2;
}

.c2 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3 {
  margin-right: 1em;
  width: 88px;
  height: 88px;
  border-radius: 36px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  border: 1px solid #4b4b4b;
}

.c3 input {
  display: none;
}

.c3 label {
  z-index: 10;
  width: 100%;
  height: 100%;
  border-radius: 36px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  text-align: center;
}

.c5 {
  width: 100%;
  border-radius: 8px;
  background-color: #494949;
  border: none;
  color: #ffff;
  padding: 0.9em;
  height: 40px;
}

.c5:focus {
  outline-width: 0;
}

.c6 {
  width: 100%;
  border-radius: 8px;
  background-color: #494949;
  border: none;
  color: #ffff;
  padding: 0.9em;
  height: 80px;
  resize: none;
}

.c6:focus {
  outline-width: 0;
}

.c7 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c9 {
  width: 98px;
  height: 41px;
  object-fit: contain;
  border-radius: 8px;
  background-color: #5f5f5f;
  color: #313131;
  border: none;
  -webkit-transition: 0.3s;
  transition: 0.3s;
  font-size: 14px;
  cursor: pointer;
}

.c9:focus {
  outline-width: 0;
}

.c9:active {
  background-color: #71bb00;
  color: #fff;
}

.c8 {
  width: 60px;
  font-size: 14px;
  color: #5f5f5f;
  background: none;
  border: none;
  cursor: pointer;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  margin-right: 1.5em;
}

.c8:focus {
  outline-width: 0;
}

.c4 {
  width: 24px;
  height: 24px;
  cursor: pointer;
  border-radius: 36px;
}

.c0 {
  padding-top: 20px;
  width: 100%;
  height: auto;
  grid-area: main;
  background-color: #343434;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c10 {
  width: 100%;
  height: 100px;
  color: #7a7a7a;
  font-size: 14px;
  grid-column-start: 2;
  grid-column-end: 2;
  display: grid;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  margin-bottom: 10px;
}

@media (min-width:900px) {
  .c1 {
    width: 100%;
  }
}

@media (min-width:900px) {
  .c0 {
    display: grid;
    grid-template-rows: auto auto auto;
    grid-template-columns: 1.5fr 2fr 1.5fr;
  }
}

@media (max-width:900px) {
  .c10 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }
}

<main
  className="c0"
>
  <form
    className="c1"
    data-test="publisher-form"
    onSubmit={[Function]}
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <label
          htmlFor="file"
        >
          <img
            alt="profile"
            className="c4"
            height="24px"
            src="image.svg"
            width="24px"
          />
        </label>
        <input
          id="file"
          name="file"
          onChange={[Function]}
          type="file"
        />
      </div>
    </div>
    <input
      className="c5"
      data-test="publisher-input-name"
      name="name"
      onChange={[Function]}
      placeholder="Digite seu nome"
      required={true}
      value=""
    />
    <textarea
      className="c6"
      data-test="publisher-textarea-msg"
      maxLength={160}
      name="message"
      onChange={[Function]}
      placeholder="Mensagem"
      required={true}
      value=""
    />
    <div
      className="c7"
    >
      <button
        className="c8"
        onClick={[Function]}
        type="reset"
      >
        Descartar
      </button>
      <button
        className="c9"
        data-test="publisher-btn-submit"
        type="submit"
      >
        Publicar
      </button>
    </div>
  </form>
  <p
    className="c10"
  >
    Feed
  </p>
</main>
`;
